extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java

```diff
@@ -411,147 +517,198 @@ public Integer get() {
-                                keyCount.increment();
-                                return result;
-                            }));
+        }
+    }
 
-                        }
-                        buffersPerPage.add(posInDictionary);
-                    }
-                }
-                List<IntBuffer> repeatCount = null;
-                if (lengthSource != null) {
-                    repeatCount = new ArrayList<>();
-                    try (final ChunkSource.GetContext context = lengthSource.makeGetContext(targetSize);
-                            final RowSequence.Iterator it = lengthRowSet.getRowSequenceIterator()) {
-                        while (it.hasMore()) {
-                            final RowSequence rs = it.getNextRowSequenceWithLength(rowStepGetter.get());
-                            // noinspection unchecked
-                            final IntChunk<Values> chunk = (IntChunk<Values>) lengthSource.getChunk(context, rs);
-                            final IntBuffer newBuffer = IntBuffer.allocate(chunk.size());
-                            chunk.copyToTypedBuffer(0, newBuffer, 0, chunk.size());
-                            newBuffer.limit(chunk.size());
-                            repeatCount.add(newBuffer);
-                        }
-                    }
-                }
-                columnWriter.addDictionaryPage(keys.getValue(), keyCount.intValue());
-                final Iterator<IntBuffer> repeatCountIt = repeatCount == null ? null : repeatCount.iterator();
-                for (final IntBuffer intBuffer : buffersPerPage) {
-                    intBuffer.flip();
-                    if (lengthSource != null) {
-                        columnWriter.addVectorPage(intBuffer, repeatCountIt.next(), intBuffer.remaining(),
-                                Integer.MIN_VALUE);
-                    } else if (hasNulls.getValue()) {
-                        columnWriter.addPage(intBuffer, Integer.MIN_VALUE, intBuffer.remaining());
+    private static <DATA_TYPE> void encodePlain(@NotNull final ParquetInstructions writeInstructions,
+            @NotNull final RowSet originalRowSet,
+            @NotNull final RowSet dataRowSet,
+            @NotNull final ColumnDefinition<DATA_TYPE> columnDefinition,
+            @NotNull final Class<DATA_TYPE> columnType,
+            @NotNull final ColumnWriter columnWriter,
+            @NotNull final ColumnSource<DATA_TYPE> dataSource,
+            @Nullable final ColumnSource<?> lengthSource,
+            @NotNull final LongSupplier rowStepGetter,
+            @NotNull final LongSupplier valuesStepGetter,
+            @NotNull final Map<String, Map<CacheTags, Object>> computedCache,
+            final int maxValuesPerPage,
+            final int maxOriginalRowsPerPage,
+            final int pageCount) throws IOException {
+        try (final TransferObject<?> transferObject = getDestinationBuffer(computedCache,
+                originalRowSet,
```

Comment:
You should notice that this is the wrong row set for columns of arrays or vectors, which will break for columns of arrays or vectors of `BigDecimal`. Not relevant to this PR, but relevant to the other one you're working on.